wt/Doc/packer.man



pack(n)                Tk Built-In Commands               pack(n)


_________________________________________________________________

NAME
       pack - Geometry manager that packs around edges of cavity

SYNOPSIS
       pack option arg ?arg ...?
_________________________________________________________________


DESCRIPTION
       The pack command is used to communicate with the packer, a
       geometry manager that arranges the children of a parent by
       packing them in order around the edges of the parent.  The
       pack command can have any of several forms,  depending  on
       the option argument:

       pack slave ?slave ...? ?options?
              If the first argument to pack is a window name (any
              value starting with ``.''),  then  the  command  is
              processed in the same way as pack configure.

       pack configure slave ?slave ...? ?options?
              The  arguments  consist of the names of one or more
              slave windows followed by pairs of  arguments  that
              specify how to manage the slaves.  See ``THE PACKER
              ALGORITHM'' below for details on  how  the  options
              are  used by the packer.  The following options are
              supported:

              -after other
                     Other must the name of another window.   Use
                     its master as the master for the slaves, and
                     insert the slaves just after  other  in  the
                     packing order.

              -anchor anchor
                     Anchor  must be a valid anchor position such
                     as n or sw; it specifies where  to  position
                     each  slave in its parcel.  Defaults to cen-
                     ter.

              -before other
                     Other must the name of another window.   Use
                     its master as the master for the slaves, and
                     insert the slaves just before other  in  the
                     packing order.

              -expand boolean
                     Specifies   whether  the  slaves  should  be
                     expanded to consume  extra  space  in  their
                     master.  Boolean may have any proper boolean
                     value, such as 1 or no.  Defaults to 0.




Tk                             4.0                              1





pack(n)                Tk Built-In Commands               pack(n)


              -fill style
                     If a  slave's  parcel  is  larger  than  its
                     requested  dimensions,  this  option  may be
                     used to stretch the slave.  Style must  have
                     one of the following values:

                      none   Give  the slave its requested dimen-
                             sions  plus  any  internal   padding
                             requested  with  -ipadx  or  -ipady.
                             This is the default.

                      x      Stretch the  slave  horizontally  to
                             fill  the entire width of its parcel
                             (except leave  external  padding  as
                             specified by -padx).

                      y      Stretch the slave vertically to fill
                             the  entire  height  of  its  parcel
                             (except  leave  external  padding as
                             specified by -pady).

                      both   Stretch the slave both  horizontally
                             and vertically.

              -in other
                     Insert  the slave(s) at the end of the pack-
                     ing order for the  master  window  given  by
                     other.

              -ipadx amount
                     Amount  specifies how much horizontal inter-
                     nal padding to leave on  each  side  of  the
                     slave(s).   Amount  must  be  a valid screen
                     distance, such as 2 or .5c.  It defaults  to
                     0.

              -ipady amount
                     Amount  specifies how much vertical internal
                     padding  to  leave  on  each  side  of   the
                     slave(s).  Amount  defaults to 0.

              -padx amount
                     Amount  specifies how much horizontal exter-
                     nal padding to leave on  each  side  of  the
                     slave(s).  Amount defaults to 0.

              -pady amount
                     Amount  specifies how much vertical external
                     padding  to  leave  on  each  side  of   the
                     slave(s).  Amount defaults to 0.

              -side side
                     Specifies  which  side  of  the  master  the
                     slave(s) will be packed  against.   Must  be



Tk                             4.0                              2





pack(n)                Tk Built-In Commands               pack(n)


                     left,  right,  top,  or bottom.  Defaults to
                     top.

              If no -in, -after or -before  option  is  specified
              then each of the slaves will be inserted at the end
              of the packing list for its  parent  unless  it  is
              already  managed  by  the  packer (in which case it
              will be left where it is).  If one of these options
              is  specified  then all the slaves will be inserted
              at the specified point.  If any of the  slaves  are
              already  managed  by  the geometry manager then any
              unspecified options for them retain their  previous
              values rather than receiving default values.

       pack forget slave ?slave ...?
              Removes  each  of the slaves from the packing order
              for its  master  and  unmaps  their  windows.   The
              slaves will no longer be managed by the packer.

       pack info slave
              Returns  a list whose elements are the current con- |
              figuration state of the slave given by slave in the |
              same  option-value  form that might be specified to |
              pack configure.  The first two elements of the list |
              are ``-in master'' where master is the slave's mas- |
              ter.

       pack propagate master ?boolean?
              If boolean has a true boolean value such as 1 or on
              then  propagation is enabled for master, which must
              be a  window  name  (see  ``GEOMETRY  PROPAGATION''
              below).   If boolean has a false boolean value then
              propagation is disabled for master.  In  either  of
              these  cases  an  empty  string  is  returned.   If
              boolean is omitted then the command returns 0 or  1
              to   indicate   whether  propagation  is  currently
              enabled for  master.   Propagation  is  enabled  by
              default.

       pack slaves master
              Returns  a list of all of the slaves in the packing
              order for master.  The order of the slaves  in  the
              list  is  the  same  as  their order in the packing
              order.  If master  has  no  slaves  then  an  empty
              string is returned.


THE PACKER ALGORITHM
       For  each  master  the packer maintains an ordered list of
       slaves called the packing  list.   The  -in,  -after,  and
       -before configuration options are used to specify the mas-
       ter for each slave and the slave's position in the packing
       list.   If none of these options is given for a slave then
       the slave is added to the end of the packing list for  its



Tk                             4.0                              3





pack(n)                Tk Built-In Commands               pack(n)


       parent.

       The  packer  arranges  the slaves for a master by scanning
       the packing list in order.  At the time it processes  each
       slave, a rectangular area within the master is still unal-
       located.  This area is called the cavity;  for  the  first
       slave it is the entire area of the master.

       For each slave the packer carries out the following steps:

       [1]    The packer allocates a rectangular parcel  for  the
              slave  along  the  side  of the cavity given by the
              slave's -side option.  If the side is top or bottom
              then  the  width  of the parcel is the width of the
              cavity and its height is the  requested  height  of
              the  slave  plus the -ipady and -pady options.  For
              the left or right side the height of the parcel  is
              the  height  of  the  cavity  and  the width is the
              requested width of the slave plus  the  -ipadx  and
              -padx  options.  The parcel may be enlarged further
              because of the -expand  option  (see  ``EXPANSION''
              below)

       [2]    The  packer  chooses  the  dimensions of the slave.
              The width will normally be  the  slave's  requested
              width  plus  twice its -ipadx option and the height
              will normally be the slave's requested height  plus
              twice  its  -ipady  option.   However, if the -fill
              option is x or both then the width of the slave  is
              expanded  to  fill  the  width of the parcel, minus
              twice the -padx option.  If the -fill option  is  y
              or both then the height of the slave is expanded to
              fill the width of the parcel, minus twice the -pady
              option.

       [3]    The packer positions the slave over its parcel.  If
              the slave is  smaller  than  the  parcel  then  the
              -anchor  option  determines where in the parcel the
              slave will be placed.  If -padx or  -pady  is  non-
              zero,  then  the  given  amount of external padding
              will always be left between the slave and the edges
              of the parcel.

       Once a given slave has been packed, the area of its parcel
       is subtracted from the cavity, leaving a smaller rectangu-
       lar cavity for the next slave.  If a slave doesn't use all
       of its parcel, the unused space in the parcel will not  be
       used  by  subsequent  slaves.  If the cavity should become
       too small to meet the needs of a slave then the slave will
       be  given  whatever  space  is left in the cavity.  If the
       cavity shrinks to zero size, then all remaining slaves  on
       the  packing  list  will be unmapped from the screen until
       the master window becomes large enough to hold them again.




Tk                             4.0                              4





pack(n)                Tk Built-In Commands               pack(n)


EXPANSION
       If  a  master  window is so large that there will be extra
       space left over after all of its slaves have been  packed,
       then the extra space is distributed uniformly among all of
       the slaves for which the -expand  option  is  set.   Extra
       horizontal  space  is  distributed  among  the  expandable
       slaves whose -side is left or right,  and  extra  vertical
       space  is  distributed  among  the expandable slaves whose
       -side is top or bottom.


GEOMETRY PROPAGATION
       The packer normally computes how large a master must be to
       just exactly meet the needs of its slaves, and it sets the
       requested width and height of the master to  these  dimen-
       sions.   This  causes geometry information to propagate up
       through a window hierarchy to a top-level window  so  that
       the  entire  sub-tree sizes itself to fit the needs of the
       leaf windows.  However, the pack propagate command may  be
       used  to turn off propagation for one or more masters.  If
       propagation is disabled then the packer will not  set  the
       requested  width  and  height  of the packer.  This may be
       useful if, for example, you wish for a  master  window  to
       have a fixed size that you specify.


RESTRICTIONS ON MASTER WINDOWS
       The  master for each slave must either be the slave's par-
       ent (the default) or a descendant of the  slave's  parent.
       This  restriction is necessary to guarantee that the slave
       can be placed over any part of its master that is  visible
       without danger of the slave being clipped by its parent.


PACKING ORDER
       If  the master for a slave is not its parent then you must
       make sure that the slave is higher in the  stacking  order
       than  the  master.   Otherwise the master will obscure the
       slave and it will appear  as  if  the  slave  hasn't  been
       packed  correctly.  The easiest way to make sure the slave
       is higher than the master is to create the  master  window
       first:   the  most recently created window will be highest
       in the stacking order.  Or, you  can  use  the  raise  and
       lower  commands to change the stacking order of either the
       master or the slave.


KEYWORDS
       geometry manager, location, packer,  parcel,  propagation,
       size







Tk                             4.0                              5




wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


wt/Wt/packer.c
/*
 * this file is part of "The W Toolkit".
 *
 * (W) 1996, Kay Roemer.
 *
 * $Id: packer.c,v 1.1.1.1 1998/11/01 19:15:04 eero Exp $
 *
 * Tk-like packer widget (written from scratch). Supports the following
 * options (the first value is the default):
 *	-fill	{none,x,y,both}
 *	-expand	{0,1}
 *	-side	{top,bottom,left,right}
 *	-anchor	{center,n,ne,e,se,s,sw,w,nw}
 *	-padx	<padding in horizontal direction, default 0>
 *	-pady	<padding in vertical direction, default 0>
 *	-ipadx	<child internal padding in horizontal direction, default 0>
 *	-ipady	<child internal padding in vertical direction, default 0>
 *
 * The widgets are packed in the order in which they are specified using the
 * WT_PACK_WIDGET option. The WT_PACK_WIDGET config option marks widgets.
 * The WT_PACK_INFO options specifies the packing configuration for the
 * widgets that are currently marked unmarks them.
 *
 * Assuming you have created a packer widget `pack' which has three childs --
 * the buttons `b1', `b2' and `b3' -- the following example creates the
 * following layout:
 *
 * +--------------------------+
 * |+----------+ +-----------+|
 * ||    b2    | |    b3     ||
 * |+----------+ +-----------+|
 * |+------------------------+|
 * ||          b1            ||
 * |+------------------------+|
 * +--------------------------+
 *
 * wt_setopt (pack,
 *	WT_PACK_WIDGET, b1,
 *	WT_PACK_INFO,   "-side bottom -fill x",
 *	WT_PACK_WIDGET, b2,
 *	WT_PACK_WIDGET, b3,
 *	WT_PACK_INFO,   "-side left",
 *	WT_EOL);
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <Wlib.h>
#include "Wt.h"
#include "toolkit.h"


typedef enum {
	FillNone,
	FillX,
	FillY,
	FillBoth
} fill_style_t;

typedef enum {
	AnchorCenter,
	AnchorN,
	AnchorNE,
	AnchorE,
	AnchorSE,
	AnchorS,
	AnchorSW,
	AnchorW,
	AnchorNW
} anchor_type_t;

typedef enum {
	SideTop,
	SideBottom,
	SideRight,
	SideLeft
} side_t;

#define IS_VERT(s)	((s) == SideTop  || (s) == SideBottom)
#define IS_HORZ(s)	(!IS_VERT (s))

typedef struct {
	long x, y, w, h;
} rect_t;

typedef struct {
	fill_style_t	fill;
	anchor_type_t	anchor;
	side_t		side;
	long		padx, pady;
	long		ipadx, ipady;
	short		expand;
	rect_t		parcel;
	long		wd, ht;
} packer_info_t;


typedef struct {
	widget_t w;
	short is_open;
	short is_realized;

	long needed_wd, needed_ht;
	short usrwd, usrht;
} packer_widget_t;

/***********************************************************************/

static inline char *
skipspace (char *cp)
{
	while (*cp && isspace (*cp))
		++cp;
	return cp;
}

static inline char *
skipword (char *cp)
{
	while (*cp && isspace (*cp))
		++cp;
	while (*cp && !isspace (*cp))
		++cp;
	while (*cp && isspace (*cp))
		++cp;
	return cp;
}

#define SEQ(s1,s2)	(!strncasecmp (s1, s2, strlen (s2)))

static packer_info_t *
parse_packer_info (char *s)
{
	packer_info_t *p;
	char *cp;

	p = malloc (sizeof (packer_info_t));
	if (!p)
		return NULL;
	memset (p, 0, sizeof (packer_info_t));

	p->fill = FillNone;
	p->side = SideTop;
	p->anchor = AnchorCenter;

	if (!s)
		return p;

	for (cp = skipspace (s); *cp; ) {
		if (SEQ (cp, "-anchor")) {
			cp = skipword (cp);
			if (SEQ (cp, "center")) {
				p->anchor = AnchorCenter;
			} else if (SEQ (cp, "ne")) {
				p->anchor = AnchorNE;
			} else if (SEQ (cp, "n")) {
				p->anchor = AnchorN;
			} else if (SEQ (cp, "se")) {
				p->anchor = AnchorSE;
			} else if (SEQ (cp, "e")) {
				p->anchor = AnchorE;
			} else if (SEQ (cp, "sw")) {
				p->anchor = AnchorSW;
			} else if (SEQ (cp, "s")) {
				p->anchor = AnchorS;
			} else if (SEQ (cp, "nw")) {
				p->anchor = AnchorNW;
			} else if (SEQ (cp, "w")) {
				p->anchor = AnchorW;
			}
		} else if (SEQ (cp, "-side")) {
			cp = skipword (cp);
			if (SEQ (cp, "top")) {
				p->side = SideTop;
			} else if (SEQ (cp, "bottom")) {
				p->side = SideBottom;
			} else if (SEQ (cp, "left")) {
				p->side = SideLeft;
			} else if (SEQ (cp, "right")) {
				p->side = SideRight;
			}
		} else if (SEQ (cp, "-fill")) {
			cp = skipword (cp);
			if (SEQ (cp, "none")) {
				p->fill = FillNone;
			} else if (SEQ (cp, "x")) {
				p->fill = FillX;
			} else if (SEQ (cp, "y")) {
				p->fill = FillY;
			} else if (SEQ (cp, "both")) {
				p->fill = FillBoth;
			}
		} else if (SEQ (cp, "-expand")) {
			cp = skipword (cp);
			p->expand = atol (cp);
		} else if (SEQ (cp, "-padx")) {
			cp = skipword (cp);
			p->padx = atol (cp);
		} else if (SEQ (cp, "-pady")) {
			cp = skipword (cp);
			p->pady = atol (cp);
		} else if (SEQ (cp, "-ipadx")) {
			cp = skipword (cp);
			p->ipadx = atol (cp);
		} else if (SEQ (cp, "-ipady")) {
			cp = skipword (cp);
			p->ipady = atol (cp);
		}
		do {
			cp = skipword (cp);
		} while (*cp && *cp != '-');
	}
	return p;
}

static packer_info_t *
clone_packer_info (packer_info_t *p)
{
	packer_info_t *newp = malloc (sizeof (packer_info_t));
	if (!newp)
		return NULL;
	*newp = *p;
	return newp;
}

static void
free_packer_info (widget_t *w)
{
	if (w->geom_info && w->geom_info != (void *)1) {
		free (w->geom_info);
		w->geom_info = NULL;
	}
}

/***********************************************************************/

/*
 * increment `space' by the amount of space needed for widget `w'
 */
static inline void
get_space_one (rect_t *space, widget_t *w)
{
	packer_info_t *p = w->geom_info;
	long wd, ht;

	wt_minsize (w, &p->wd, &p->ht);
	wd = p->wd + 2*(p->padx + p->ipadx);
	ht = p->ht + 2*(p->pady + p->ipady);

	switch (p->side) {
	case SideTop:
	case SideBottom:
		space->h += ht;
		if (space->w < wd)
			space->w = wd;
		break;

	case SideLeft:
	case SideRight:
		space->w += wd;
		if (space->h < ht)
			space->h = ht;
		break;
	}
}

/*
 * get the minimal space needed for packing the widgets pointed at by 'w'.
 */
static int
get_space (widget_t *w, long *wd, long *ht)
{
	widget_t *wp;
	rect_t space;

	if (!w) {
		*wd = 0;
		*ht = 0;
		return 0;
	}
	for (wp = w; wp->next; wp = wp->next)
		;
	space.w = 0;
	space.h = 0;
	for ( ;; wp = wp->prev) {
		if (!wp->geom_info || wp->geom_info == (void *)1) {
			wp->geom_info = parse_packer_info (NULL);
			if (!wp->geom_info)
				return -1;
		}
		get_space_one (&space, wp);
		if (wp == w)
			break;
	}
	*wd = space.w;
	*ht = space.h;
	return 0;
}

/*
 * get the amount of space by which the widgets parcel must be expanded
 * to fill the remaining space in the packer window.
 */
static inline long
get_expand_space (rect_t *space, widget_t *wp)
{
	long isvert, i, exp_space;
	packer_info_t *p;

	p = wp->geom_info;
	isvert = IS_VERT (p->side);
	exp_space = isvert ? space->h : space->w;
	for (i = 0; wp; wp = wp->next) {
		p = wp->geom_info;
		if (isvert ^ IS_VERT (p->side))
			break;
		if (p->expand)
			++i;
		exp_space -= isvert ? p->ht + 2*(p->ipady + p->pady)
				    : p->wd + 2*(p->ipadx + p->padx);
	}
	if (wp && i > 0) {
		long wd, ht;
		get_space (wp, &wd, &ht);
		exp_space -= isvert ? ht : wd;
	}
	return (i > 0 && exp_space > 0) ? exp_space/i : 0;
}

/*
 * allocate a parcel for widget `w' and reduce `space' by the space needed
 * for the parcel.
 */
static void
get_parcel (rect_t *space, widget_t *w)
{
	packer_info_t *p = w->geom_info;
	long wd, ht, exp_space;

	wd = p->wd + 2*(p->padx + p->ipadx);
	ht = p->ht + 2*(p->pady + p->ipady);

	exp_space = p->expand ? get_expand_space (space, w) : 0;

	switch (p->side) {
	case SideTop:
		ht += exp_space;
		p->parcel.x = space->x;
		p->parcel.y = space->y;
		p->parcel.w = MAX (wd, space->w);
		p->parcel.h = ht;
		space->y   += ht;
		space->h   -= ht;
		break;

	case SideBottom:
		ht += exp_space;
		p->parcel.x = space->x;
		p->parcel.y = space->y + space->h - ht;
		p->parcel.w = MAX (wd, space->w);
		p->parcel.h = ht;
		space->h   -= ht;
		break;

	case SideLeft:
		wd += exp_space;
		p->parcel.x = space->x;
		p->parcel.y = space->y;
		p->parcel.w = wd;
		p->parcel.h = MAX (ht, space->h);
		space->x   += wd;
		space->w   -= wd;
		break;

	case SideRight:
		wd += exp_space;
		p->parcel.x = space->x + space->w - wd;
		p->parcel.y = space->y;
		p->parcel.w = wd;
		p->parcel.h = MAX (ht, space->h);
		space->w   -= wd;
		break;
	}
}

/*
 * place a widget within its parcel
 */
static void
place_widget (widget_t *w)
{
	packer_info_t *p = w->geom_info;
	long x, y, wd, ht;

	wd = (p->fill == FillX || p->fill == FillBoth)
		? p->parcel.w - 2*p->padx
		: p->wd + 2*p->ipadx;

	ht = (p->fill == FillY || p->fill == FillBoth)
		? p->parcel.h - 2*p->pady
		: p->ht + 2*p->ipady;

	wt_reshape (w, WT_UNSPEC, WT_UNSPEC, wd, ht);
	/*
	 * the widget might not be resizeable, so ask again
	 */
	wt_geometry (w, &x, &y, &wd, &ht);

	switch (p->anchor) {
	case AnchorSW:
	case AnchorW:
	case AnchorNW:
		x = p->parcel.x + p->padx;
		break;
	case AnchorN:
	case AnchorCenter:
	case AnchorS:
		x = p->parcel.x + (p->parcel.w - wd)/2;
		break;
	case AnchorNE:
	case AnchorE:
	case AnchorSE:
		x = p->parcel.x + p->parcel.w - wd - p->padx;
		break;
	}
	switch (p->anchor) {
	case AnchorNW:
	case AnchorN:
	case AnchorNE:
		y = p->parcel.y + p->pady;
		break;
	case AnchorW:
	case AnchorCenter:
	case AnchorE:
		y = p->parcel.y + (p->parcel.h - ht)/2;
		break;
	case AnchorSW:
	case AnchorS:
	case AnchorSE:
		y = p->parcel.y + p->parcel.h - ht - p->pady;
		break;
	}
	wt_reshape (w, x, y, WT_UNSPEC, WT_UNSPEC);
}

/*
 * place all the child widgets
 */
static int
pack_widgets (packer_widget_t *w, int firsttime)
{
	widget_t *wp;
	WWIN *win = w->w.win;
	rect_t space;

	if (!w->w.childs)
		return 0;

	if (get_space (w->w.childs, &w->needed_wd, &w->needed_ht) < 0)
		return -1;

	space.x = 0;
	space.y = 0;
	space.w = MAX (w->w.w, w->needed_wd);
	space.h = MAX (w->w.h, w->needed_ht);

	for (wp = w->w.childs; wp; wp = wp->next) {
		get_parcel (&space, wp);
		place_widget (wp);
		if (firsttime && (*wp->class->realize) (wp, win) < 0)
			return -1;
	}
	return 0;
}

/***********************************************************************/

static long packer_query_geometry (widget_t *, long *, long *, long *, long *);

static long
packer_init (void)
{
	return 0;
}

static widget_t *
packer_create (widget_class_t *cp)
{
	packer_widget_t *wp = malloc (sizeof (packer_widget_t));
	if (!wp)
		return NULL;
	memset (wp, 0, sizeof (packer_widget_t));
	wp->w.class = wt_packer_class;
	return (widget_t *)wp;
}

static long
packer_delete (widget_t *_w)
{
	packer_widget_t *w = (packer_widget_t *)_w;
	widget_t *wp, *next;

	for (wp = w->w.childs; wp; wp = next) {
		next = wp->next;
		free_packer_info (wp);
		(*wp->class->delete) (wp);
	}
	if (w->is_realized)
		w_delete (w->w.win);
	free (w);
	return 0;
}

static long
packer_close (widget_t *_w)
{
	packer_widget_t *w = (packer_widget_t *)_w;

	if (w->is_realized && w->is_open) {
		w_close (w->w.win);
		w->is_open = 0;
	}
	return 0;
}

static long
packer_open (widget_t *_w)
{
	packer_widget_t *w = (packer_widget_t *)_w;

	if (w->is_realized && !w->is_open) {
		w_open (w->w.win, w->w.x, w->w.y);
		w->is_open = 1;
	}
	return 0;
}

static long
packer_addchild (widget_t *parent, widget_t *w)
{
	widget_t *wp;

	for (wp = parent->childs; wp && wp->next; wp = wp->next)
		;
	wt_add_after (parent, wp, w);
	return 0;
}

static long
packer_delchild (widget_t *parent, widget_t *w)
{
	free_packer_info (w);
	wt_remove (w);
	return 0;
}

static long
packer_realize (widget_t *_w, WWIN *parent)
{
	packer_widget_t *w = (packer_widget_t *)_w;
	long x, y, wd, ht;

	if (w->is_realized)
		return -1;

	packer_query_geometry (_w, &x, &y, &wd, &ht);
	w->w.w = wd;
	w->w.h = ht;

	w->w.win = wt_create_window (parent, wd, ht,
		W_MOVE|W_CONTAINER|W_NOBORDER, _w);
	if (!w->w.win)
		return -1;
	w->w.win->user_val = (long)w;
	if (pack_widgets (w, 1) < 0) {
		w_delete (w->w.win);
		return -1;
	}
	w->is_realized = 1;
	w->is_open = 1;
	w_open (w->w.win, w->w.x, w->w.y);
	return 0;
}

static long
packer_query_geometry (widget_t *_w, long *xp, long *yp, long *wdp, long *htp)
{
	packer_widget_t *w = (packer_widget_t *)_w;

	*xp = w->w.x;
	*yp = w->w.y;

	if (w->w.w > 0 && w->w.h > 0) {
		*wdp = w->w.w;
		*htp = w->w.h;
	} else {
		(*_w->class->query_minsize) (_w, wdp, htp);
	}
	if (*wdp <= 0)
		*wdp = 10;
	if (*htp <= 0)
		*htp = 10;
	return 0;
}

static long
packer_query_minsize (widget_t *_w, long *wdp, long *htp)
{
	packer_widget_t *w = (packer_widget_t *)_w;

	get_space (_w->childs, wdp, htp);
	*wdp = MAX (*wdp, w->usrwd);
	*htp = MAX (*htp, w->usrht);
	return 0;
}

static long
packer_reshape (widget_t *_w, long x, long y, long wd, long ht)
{
	packer_widget_t *w = (packer_widget_t *)_w;
	int ret = 0;

	if (x != w->w.x || y != w->w.y) {
		if (w->is_realized)
			w_move (w->w.win, x, y);
		w->w.x = x;
		w->w.y = y;
		ret = 1;
	}
	if (wd != w->w.w || ht != w->w.h) {
		w->w.w = wd;
		w->w.h = ht;
		if (w->is_realized) {
			w_resize (w->w.win, wd, ht);
			pack_widgets (w, 0);
		}
		ret = 1;
	}
	return ret;
}

static int
add_managed_child (packer_widget_t *w, widget_t *child)
{
	int not_managed = 0;
	widget_t *wp, *prev;

	for (wp = w->w.childs; wp; wp = wp->next) {
		if (wp == child) {
			if (!child->geom_info) {
				wt_remove (child);
				not_managed = 1;
			}
			free_packer_info (wp);
			wp->geom_info = (void *)1;
			break;
		}
	}
	if (!wp) {
		return -1;
	}
	if (not_managed) {
		prev = NULL;
		for (wp = w->w.childs; wp && wp->geom_info; ) {
			prev = wp;
			wp = wp->next;
		}
		if (prev) {
			wt_add_after ((widget_t *)w, prev, child);
		} else {
			wt_add_before ((widget_t *)w, w->w.childs, child);
		}
	}
	return 0;
}

static long
packer_setopt (widget_t *_w, long key, void *val)
{
	packer_widget_t *w = (packer_widget_t *)_w;
	packer_info_t *p;
	widget_t *wp;
	short mask = 0;
	int i;

	switch (key) {
	case WT_XPOS:
		if (packer_reshape (_w, *(long *)val, w->w.y, w->w.w, w->w.h))
			mask |= WT_CHANGED_POS;
		break;

	case WT_YPOS:
		if (packer_reshape (_w, w->w.x, *(long *)val, w->w.w, w->w.h))
			mask |= WT_CHANGED_POS;
		break;

	case WT_WIDTH:
		w->usrwd = MAX (0, *(long *)val);
		if (packer_reshape (_w, w->w.x, w->w.y, w->usrwd, w->w.h))
			mask |= WT_CHANGED_SIZE;
		break;

	case WT_HEIGHT:
		w->usrht = MAX (0, *(long *)val);
		if (packer_reshape (_w, w->w.x, w->w.y, w->w.w, w->usrht))
			mask |= WT_CHANGED_SIZE;
		break;

	case WT_PACK_WIDGET:
		if (add_managed_child (w, val) < 0)
			return -1;
		break;

	case WT_PACK_INFO:
		p = parse_packer_info (val);
		if (!p)
			return -1;
		for (i = 0, wp = w->w.childs; wp; wp = wp->next) {
			if (wp->geom_info == (void *)1) {
				if (i == 0) {
					i = 1;
					wp->geom_info = p;
				} else {
					wp->geom_info = clone_packer_info (p);
					if (!wp->geom_info)
						break;
				}
			}
		}
		if (i == 0)
			free (p);
		if (wp != NULL)
			return -1;
		break;

	default:
		return -1;
	}
	if (mask && w->is_realized)
		wt_change_notify (_w, mask);
	return 0;
}

static long
packer_getopt (widget_t *_w, long key, void *val)
{
	packer_widget_t *w = (packer_widget_t *)_w;

	switch (key) {
	case WT_XPOS:
		*(long *)val = w->w.x;
		break;

	case WT_YPOS:
		*(long *)val = w->w.y;
		break;

	case WT_WIDTH:
		*(long *)val = w->w.w;
		break;

	case WT_HEIGHT:
		*(long *)val = w->w.h;
		break;

	default:
		return -1;
	}
	return 0;
}

static WEVENT *
packer_event (widget_t *_w, WEVENT *ev)
{
	return ev;
}

static long
packer_changes (widget_t *w, widget_t *w2, short changes)
{
	return 0;
}

static widget_class_t _wt_packer_class = {
	"packer", 0,
	packer_init,
	packer_create,
	packer_delete,
	packer_close,
	packer_open,
	packer_addchild,
	packer_delchild,
	packer_realize,
	packer_query_geometry,
	packer_query_minsize,
	packer_reshape,
	packer_setopt,
	packer_getopt,
	packer_event,
	packer_changes,
	packer_changes
};

widget_class_t *wt_packer_class = &_wt_packer_class;
